### ISSUE TITLE
Users can easily impersonate other users

### ISSUE THREAD

**Issue by ayushin** [chatwoot/chatwoot#2762]
**Describe the bug**

Any user can impersonate any user by supplying their email / name to the javascript chat widget via setUser(...

**To Reproduce**

Call `setUser` without any `identifier_hash` for any existing user and start the chat. The conversation will be assigned to the exiting user and there is no way to tell this is not the original user you are talking to.

**Expected behavior**

Do not merge unverified and verified users by user identifier. Clearly mark those identified and unidentified users.

**Comment by pranavrajs** [comment]
Please enable identity validation if you want to disable this behavior. https://www.chatwoot.com/docs/product/channels/live-chat/sdk/setup#identity-validation-using-hmac

**Comment by pranavrajs** [comment]
We can think of showing the session as verified as described in #1682

**Comment by pranavrajs** [comment]
I'm closing this for now, please feel free to re-open if you don't feel that #1682 solves the problem.

**Comment by ayushin** [comment]
@pranavrajs Thanks for your response, just can't seem where I can enable / enforce identity check in the admin on chatwoot.com?

**Comment by hodl** [comment]
Having the same issue, even when setting the `identifier_hash`.
Additionally I decided I would try setting a custom attribute of the `identifier_hash` using `setCustomAttributes` and while it initially works as expected as soon as the "malicious" user provides the spoofed email address the custom attribute changes to the `identifier_hash` of the "innocent" user!

**Comment by pranavrajs** [comment]
@ayushin @hodl Makes sense, we will add a toggle on the inbox settings to enforce the identification. So that if the identifier_hash is not available the call will fail.

**Comment by ayushin** [comment]
I think the next thing to consider is use a standard JWT for passing user details 

https://github.com/chatwoot/chatwoot/issues/2745

**Comment by WevrLabs** [comment]
> @ayushin @hodl Makes sense, we will add a toggle on the inbox settings to enforce the identification. So that if the identifier_hash is not available the call will fail.

Besides failing, it could ask the user to login to thier account with X company, just to provide the user with a meaningful UX. 

It could be something like this:

> Please login to your account on X Company, to be able to initiate a chat session (https://xcompany.com/login)

Of course, `X Company` and the url are customizable on the inbox settings. 

Moreover, this should only be triggered in case the passed email exists previously in the database, but only if its contact identifier_hash was there, i.e the email was provided but without the identifier_hash and that in a previous time, it had an identifier_hash passed with it. This is to avoid preventing chats from visitors who initiate a conversation for the first time,  and not yet signed up an account with X Company. 

This can be optional too, in other words, chats can be completely prevented if no identifier_hash was passed in sdk, which is very useful in case scenarios where chatbox needs to be displayed only to logged in users.

**Comment by hodl** [comment]
Just keep in mind us that do not have user logins/accounts. I am creating a custom identifier set as a session cookie, and hashing that identifier. We hope the user will provide email address for best support results, but it isn't always the case. `identifier_hash` needs to be properly applied even if user doesn't provide an email address initially. 

We also do not need the call to fail, we just need to see reliably in the UI that conversations between these two entities are not shared. Maybe they are the same person, maybe not, they shouldn't be able to see conversation history unless the `identifier_hash` is matching.

**Comment by pranavrajs** [comment]
@fayazara Please take a look at this issue and suggest if you need any more input on this.

**Comment by ash-auctionvilla** [comment]
Sorry if this issue has been resolved already. But does #3137 solves this issue as @hodl suggested above? It seems to enforce validation in inboxes which is a problem itself.

**Comment by ash-auctionvilla** [comment]
> @ash-auctionvilla Could you please share a use case or scenario, which might help us better understand your issue.

The use case - identifying if users are already authenticated so that we may provide them account related/sensitive services. 

i.e. a user places an order, realize they entered incorrect address and wants to change through website live chat. The agents should be able to see a verified flag as explained in #1682 - making sure no bad actor is impersonating the user.
